### ISSUE TITLE
Transient Failures Causing CancellationToken to be cancelled instead of retry after upgrade to 6.0

### ISSUE THREAD

**Issue by m-hicks-OH** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1727]
### Steps to reproduce
We have recently upgraded our code base from .net core 3.1 to .net 6.0 and as part of that change upgraded EntityFrameworkCore and Pomelo.  Additionally, it seems we've moved from using `MySql.Data.MySqlClient` to `MySqlConnector`.

We have an ETL process that runs multiple threads for a job in parallel and calls `ExecuteRawSqlAsync` against our database stored procedures.  What we started seeing with 6.0 is that we're getting cancellationToken cancellations and we can't figure out why.  Currently, I'm thinking there was a behavior change with the way transient failures are handled if not retried.  It seems as though any transient error we get results in the parent cancellationToken getting cancelled and killing the rest of the DB calls.

We are using `.EnableRetryOnFalue` explicitly on the `.UseMySql()` optionbuilder, but it seems that `ExecuteRawSqlAsync` doesn't use the current execution strategy.  What's odd is that this worked before in 3.1.  Additionally, we're trying Polly retries, but I think they may be retrying once because it detects the failure, but it won't succeed on retry because at that point the token is cancelled.

### The issue
Basically, we're seeing transient errors result in tokens being cancelled instead of the connection retried.

### Further technical details
We've had a handful of our team members digging into this issue as we're trying to get off of 3.1 before it leaves support.

One thing I'm seeing might have been fixed in MySqlConnector 2.1.13, but it appears our version of Pomelo is still on 2.1.12.

MySQL version: 5.6
Pomelo.EntityFrameworkCore.MySql version: 6.0.2
Microsoft.AspNetCore.App version: 6.0.2

**Comment by lauxjpn** [comment]
> We are using .EnableRetryOnFalue explicitly on the .UseMySql() optionbuilder, but it seems that ExecuteRawSqlAsync doesn't use the current execution strategy.

That is correct. From [ExecuteSqlRawAsync(DatabaseFacade, String, CancellationToken): Remarks]:(https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationaldatabasefacadeextensions.executesqlrawasync?view=efcore-6.0#microsoft-entityframeworkcore-relationaldatabasefacadeextensions-executesqlrawasync(microsoft-entityframeworkcore-infrastructure-databasefacade-system-string-system-threading-cancellationtoken))

> ### Remarks
> Note that this method does not start a transaction. To use this method with a transaction, first call [BeginTransaction(DatabaseFacade, IsolationLevel)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationaldatabasefacadeextensions.begintransaction?view=efcore-6.0#microsoft-entityframeworkcore-relationaldatabasefacadeextensions-begintransaction(microsoft-entityframeworkcore-infrastructure-databasefacade-system-data-isolationlevel)) or UseTransaction.
>
> **Note that the current [ExecutionStrategy](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.storage.executionstrategy?view=efcore-6.0) is not used by this method since the SQL may not be idempotent and does not run in a transaction. An [ExecutionStrategy](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.storage.executionstrategy?view=efcore-6.0) can be used explicitly, making sure to also use a transaction if the SQL is not idempotent.**
>
> Never pass a concatenated or interpolated string ($"") with non-validated user-provided values into this method. Doing so may expose your application to SQL injection attacks.
>
> See [Executing raw SQL commands with EF Core](https://aka.ms/efcore-docs-raw-sql) for more information and examples.

---

> One thing I'm seeing might have been fixed in MySqlConnector 2.1.13, but it appears our version of Pomel is still on 2.1.12.

You can just update the MySqlConnector package to a higher version.

**Comment by m-hicks-OH** [comment]
Thanks for the quick reply.
So we have a reference to Pomelo but not directly to MySqlConnector in our csproj.
Do we just add MySqlConnector reference explicitly in the csproj?

**Comment by lauxjpn** [comment]
> So we have a reference to Pomelo but not directly to MySqlConnector in our csproj.
> Do we just add MySqlConnector reference explicitly in the csproj?

Yes, something like the following:

```xml
<PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="6.0.2" />
<PackageReference Include="MySqlConnector" Version="2.1.13" />
```

Also, be aware that MySQL 5.6 is 9 years old and not officially supported by Oracle (no more updates) or us anymore. It had reached its end of support in February 2021.

**Comment by m-hicks-OH** [comment]
Thanks!
Yes, our test environment was just updated to MySql 8.0 and we're in the process of getting our stage and prod off 5.6 as well.

**Comment by lauxjpn** [comment]
> An [ExecutionStrategy](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.storage.executionstrategy?view=efcore-6.0) can be used explicitly, making sure to also use a transaction if the SQL is not idempotent.

For an example on how to use an ExecutionStrategy explicitly, see [Connection Resiliency: Execution strategies and transactions](https://learn.microsoft.com/en-us/ef/core/miscellaneous/connection-resiliency#execution-strategies-and-transactions).

**Comment by m-hicks-OH** [comment]
Thanks again.  We're updating the package for MySqlConnector in our test environment as I type.
I've also wrapped one of our async sproc calls with an explicit ExecutionStrategy with retry enabled.

Fingers-crossed.  I'll post back on here with the outcome.

**Comment by m-hicks-OH** [comment]
We've got all of our updates in as well as a change to update the way our cancellationTokens are passed into our background processes.  There's a non-zero chance that that last change was the real kicker.  We're deploying our change to production tonight, as test/stage tested out with flying colors.

**Comment by m-hicks-OH** [comment]
Okay, well all the changes paid off, and it looks like we're in the clear!
Thanks again!

**Comment by lauxjpn** [comment]
@m-hicks-OH Thanks for letting us know!
